packages/rpc-core/src/default-commitment.ts

```diff
@@ -0,0 +1,60 @@
+import { Commitment } from '@solana/rpc-types';
+
+type Config = Readonly<{
+    params: unknown[];
+    commitmentPropertyName: string;
+    optionsObjectPositionInParams: number;
+    overrideCommitment?: Commitment;
+}>;
+
+export function applyDefaultCommitment({
+    commitmentPropertyName,
+    params,
+    optionsObjectPositionInParams,
+    overrideCommitment,
+}: Config) {
+    const paramInTargetPosition = params[optionsObjectPositionInParams];
+    if (
+        // There's no config.
+        paramInTargetPosition === undefined ||
+        // There is a config object.
+        (paramInTargetPosition && typeof paramInTargetPosition === 'object' && !Array.isArray(paramInTargetPosition))
+    ) {
+        if (
+            // The config object already has a commitment set.
+            paramInTargetPosition &&
+            commitmentPropertyName in paramInTargetPosition
+        ) {
+            if (
+                !paramInTargetPosition[commitmentPropertyName as keyof typeof paramInTargetPosition] ||
+                paramInTargetPosition[commitmentPropertyName as keyof typeof paramInTargetPosition] === 'finalized'
+            ) {
+                // Delete the commitment property; `finalized` is already the server default.
+                const nextParams = [...params];
+                const {
+                    [commitmentPropertyName as keyof typeof paramInTargetPosition]: _, // eslint-disable-line @typescript-eslint/no-unused-vars
+                    ...rest
+                } = paramInTargetPosition;
+                if (Object.keys(rest).length > 0) {
+                    nextParams[optionsObjectPositionInParams] = rest;
+                } else {
+                    if (optionsObjectPositionInParams === nextParams.length - 1) {
+                        nextParams.length--;
+                    } else {
+                        nextParams[optionsObjectPositionInParams] = undefined;
+                    }
+                }
+                return nextParams;
+            }
```

Comment:
Is this part necessary? I understand that the server default is `finalized` which means it's redundant to provide it explicitly but I'm not sure why we would take on the responsibility of removing it when we're only trying to set a better implicit default value here.

Reply:
There is a reason! Because this happens pre-transport, we have this one last opportunity to normalize the params. Doing so makes it easier down the pipe for, say, a memoizing transport. Now the memoization routine doesn't have to hash these to the same value:

* `['getBalance', 'foo', {commitment: 'finalized'}]`
* `['getBalance', 'foo', {commitment: undefined}]`
* `['getBalance', 'foo']`

…since we normalize them ahead of time.